### ISSUE TITLE
qt (qtassistant?) fails to build with gcc thread model posix

### ISSUE THREAD

**Issue by uwehermann** [mxe/mxe#565]
We're using Qt4 from MXE currently to build the sigrok / PulseView installers for Windows. We also need to use the gcc thread model "posix" (not "win32") since PulseView uses C++11 (and C++11 std::thread).

However, "make qt" fails to build when gcc's thread model is set to "posix". Apparently the qtassistant part has issues. If you comment out the snippet building that in src/qt.mk the rest of Qt seems to build fine.

```
-    cd '$(1)/tools/assistant' && '$(1)/bin/qmake' assistant.pro
-    # can't figure out where -lQtCLucene comes from so use
-    # sed on the output instead of patching the input
-    $(MAKE) -C '$(1)/tools/assistant' sub-lib-qmake_all
-    $(SED) -i 's,-lQtCLucene$$,-lQtCLucene4,' '$(1)/tools/assistant/lib/Makefile.Release'
-    $(MAKE) -C '$(1)/tools/assistant' -j '$(JOBS)' install
+    ## cd '$(1)/tools/assistant' && '$(1)/bin/qmake' assistant.pro
+    ## # can't figure out where -lQtCLucene comes from so use
+    ## # sed on the output instead of patching the input
+    ## $(MAKE) -C '$(1)/tools/assistant' sub-lib-qmake_all
+    ## $(SED) -i 's,-lQtCLucene$$,-lQtCLucene4,' '$(1)/tools/assistant/lib/Makefile.Release'
+    ## $(MAKE) -C '$(1)/tools/assistant' -j '$(JOBS)' install
```

Error log (if you don't do the commenting above):

```
i686-w64-mingw32.static-g++ -c -pipe -fno-keep-inline-dllexport -O2 -Wall -Wextra -frtti -fexceptions -mthreads -DUNICODE -DQ_USE_SYBASE -DQT_NO_DEBUG -DQT_SQL_LIB -DQT_GUI_LIB -DQT_NETWORK_LIB -DQT_CORE_LIB -DQT_HAVE_MMX -DQT_HAVE_3DNOW -DQT_HAVE_SSE -DQT_HAVE_MMXEXT -DQT_HAVE_SSE2 -DQT_HAVE_SSE3 -DQT_HAVE_SSSE3 -DQT_HAVE_SSE4_1 -DQT_HAVE_SSE4_2 -DQT_HAVE_AVX -DQT_THREAD_SUPPORT -I'../../../../include/QtCore' -I'../../../../include/QtNetwork' -I'../../../../include/QtGui' -I'../../../../include/QtSql' -I'../../../../include' -I'../../../../include/QtHelp' -I'../../../shared/fontpanel' -I'.' -I'../../../../include/ActiveQt' -I'.moc/release-static' -I'.uic/release-static' -I'../../../../mkspecs/win32-g++-4.6' -o .obj/release-static/qrc_assistant_images.o .rcc/release-static/qrc_assistant_images.cpp
i686-w64-mingw32.static-g++ -Wl,-s -mthreads -Wl,-subsystem,windows -o ../../../../bin/assistant.exe object_script.assistant.Release  -L'/home/uwe/mxe-git/tmp-qt-i686-w64-mingw32.static/qt-everywhere-opensource-src-4.8.6/lib' -L'/home/uwe/mxe-git/tmp-qt-i686-w64-mingw32.static/qt-everywhere-opensource-src-4.8.6/plugins/sqldrivers' -L'/home/uwe/mxe-git/tmp-qt-i686-w64-mingw32.static/qt-everywhere-opensource-src-4.8.6/lib' -lmingw32 -lqtmain .obj/release-static/assistant_res.o -lQtHelp -lQtSql -lQtGui -lQtNetwork -lQtCLucene -lpq -lsecur32 -lodbc32 -lsybdb -lgnutls -lhogweed -lnettle -lgmp -liconv -lsqlite3 -lcomdlg32 -loleaut32 -limm32 -lwinmm -lwinspool -lmsimg32 -lmng -llcms2 -lm -ltiff -llzma -ljpeg -lpng -lssl -lcrypto -lgdi32 -lcrypt32 -lQtCore -lole32 -luuid -lws2_32 -ladvapi32 -lshell32 -luser32 -lkernel32 -lz 
/home/uwe/mxe-git/usr/lib/gcc/i686-w64-mingw32.static/4.9.2/../../../../i686-w64-mingw32.static/lib/../lib/libpthread.a(libwinpthread_la-thread.o): In function `pthread_getspecific':
/home/uwe/mxe-git/tmp-winpthreads-i686-w64-mingw32.static/mingw-w64-v3.3.0/mingw-w64-libraries/winpthreads/src/thread.c:752: multiple definition of `pthread_getspecific'
/home/uwe/mxe-git/usr/lib/gcc/i686-w64-mingw32.static/4.9.2/../../../../i686-w64-mingw32.static/lib/../lib/libpq.a(pthread-win32.o):pthread-win32.c:(.text+0x20): first defined here
/home/uwe/mxe-git/usr/lib/gcc/i686-w64-mingw32.static/4.9.2/../../../../i686-w64-mingw32.static/lib/../lib/libpthread.a(libwinpthread_la-thread.o): In function `pthread_setspecific':
/home/uwe/mxe-git/tmp-winpthreads-i686-w64-mingw32.static/mingw-w64-v3.3.0/mingw-w64-libraries/winpthreads/src/thread.c:765: multiple definition of `pthread_setspecific'
/home/uwe/mxe-git/usr/lib/gcc/i686-w64-mingw32.static/4.9.2/../../../../i686-w64-mingw32.static/lib/../lib/libpq.a(pthread-win32.o):pthread-win32.c:(.text+0x10): first defined here
/home/uwe/mxe-git/usr/lib/gcc/i686-w64-mingw32.static/4.9.2/../../../../i686-w64-mingw32.static/lib/../lib/libpthread.a(libwinpthread_la-thread.o): In function `pthread_self':
/home/uwe/mxe-git/tmp-winpthreads-i686-w64-mingw32.static/mingw-w64-v3.3.0/mingw-w64-libraries/winpthreads/src/thread.c:912: multiple definition of `pthread_self'
/home/uwe/mxe-git/usr/lib/gcc/i686-w64-mingw32.static/4.9.2/../../../../i686-w64-mingw32.static/lib/../lib/libpq.a(pthread-win32.o):pthread-win32.c:(.text+0x0): first defined here
/home/uwe/mxe-git/usr/lib/gcc/i686-w64-mingw32.static/4.9.2/../../../../i686-w64-mingw32.static/lib/../lib/libpthread.a(libwinpthread_la-mutex.o): In function `pthread_mutex_unlock':
/home/uwe/mxe-git/tmp-winpthreads-i686-w64-mingw32.static/mingw-w64-v3.3.0/mingw-w64-libraries/winpthreads/src/mutex.c:317: multiple definition of `pthread_mutex_unlock'
/home/uwe/mxe-git/usr/lib/gcc/i686-w64-mingw32.static/4.9.2/../../../../i686-w64-mingw32.static/lib/../lib/libpq.a(pthread-win32.o):pthread-win32.c:(.text+0xa0): first defined here
/home/uwe/mxe-git/usr/lib/gcc/i686-w64-mingw32.static/4.9.2/../../../../i686-w64-mingw32.static/lib/../lib/libpthread.a(libwinpthread_la-mutex.o): In function `pthread_mutex_init':
/home/uwe/mxe-git/tmp-winpthreads-i686-w64-mingw32.static/mingw-w64-v3.3.0/mingw-w64-libraries/winpthreads/src/mutex.c:400: multiple definition of `pthread_mutex_init'
/home/uwe/mxe-git/usr/lib/gcc/i686-w64-mingw32.static/4.9.2/../../../../i686-w64-mingw32.static/lib/../lib/libpq.a(pthread-win32.o):pthread-win32.c:(.text+0x30): first defined here
/home/uwe/mxe-git/usr/lib/gcc/i686-w64-mingw32.static/4.9.2/../../../../i686-w64-mingw32.static/lib/../lib/libpthread.a(libwinpthread_la-mutex.o): In function `pthread_mutex_lock':
/home/uwe/mxe-git/tmp-winpthreads-i686-w64-mingw32.static/mingw-w64-v3.3.0/mingw-w64-libraries/winpthreads/src/mutex.c:237: multiple definition of `pthread_mutex_lock'
/home/uwe/mxe-git/usr/lib/gcc/i686-w64-mingw32.static/4.9.2/../../../../i686-w64-mingw32.static/lib/../lib/libpq.a(pthread-win32.o):pthread-win32.c:(.text+0x70): first defined here
collect2: error: ld returned 1 exit status
Makefile.Release:198: recipe for target '../../../../bin/assistant.exe' failed
make[5]: *** [../../../../bin/assistant.exe] Error 1
make[5]: Leaving directory '/home/uwe/mxe-git/tmp-qt-i686-w64-mingw32.static/qt-everywhere-opensource-src-4.8.6/tools/assistant/tools/assistant'
Makefile:46: recipe for target 'release-install' failed
make[4]: *** [release-install] Error 2
make[4]: Leaving directory '/home/uwe/mxe-git/tmp-qt-i686-w64-mingw32.static/qt-everywhere-opensource-src-4.8.6/tools/assistant/tools/assistant'
Makefile:62: recipe for target 'sub-assistant-install_subtargets-ordered' failed
make[3]: *** [sub-assistant-install_subtargets-ordered] Error 2
make[3]: Leaving directory '/home/uwe/mxe-git/tmp-qt-i686-w64-mingw32.static/qt-everywhere-opensource-src-4.8.6/tools/assistant/tools'
Makefile:133: recipe for target 'sub-tools-install_subtargets-ordered' failed
make[2]: *** [sub-tools-install_subtargets-ordered] Error 2
make[2]: Leaving directory '/home/uwe/mxe-git/tmp-qt-i686-w64-mingw32.static/qt-everywhere-opensource-src-4.8.6/tools/assistant'
Makefile:438: recipe for target 'build-only-qt_i686-w64-mingw32.static' failed
make[1]: *** [build-only-qt_i686-w64-mingw32.static] Error 2
make[1]: Leaving directory '/home/uwe/mxe-git'
```

**Comment by uwehermann** [comment]
Any updates on this issue? The build is still failing for me with git HEAD of MXE as of today.

**Comment by bleverett** [comment]
I had the same issue, but compiling with a different version of GCC fixed the problem. x86_64-pc-linux-gnu-4.8.5 was not able to compile postgresql, but x86_64-pc-linux-gnu-4.9.3 was able to compile it.  I am talking about the version of gcc that your host machine is running, not the version of gcc that you compile in MXE.

**Comment by tonytheodore** [comment]
@uwehermann should be fixed with #1009, `enable_thread_safety` really just means "build internal pthreads".

**Comment by tonytheodore** [comment]
#1009 has been reverted for now https://github.com/mxe/mxe/commit/1a94ae0e6777c841e005366bf0f85b51a3c9f959

**Comment by ByrAnd** [comment]
In my case postgresql definitions caused errors in libmysqlclient. I prepended sed renaming of corresponding functions before compilation.  This solved the problem - for now.

```
diff --git a/src/libmysqlclient.mk b/src/libmysqlclient.mk
index ee22205..8b69eb1 100644
--- a/src/libmysqlclient.mk
+++ b/src/libmysqlclient.mk
@@ -17,6 +17,25 @@ define $(PKG)_UPDATE
 endef
 
 define $(PKG)_BUILD
+$(info PREFIX/TARGET is [${PREFIX}]/[${TARGET}])
+    # http://stackoverflow.com/questions/35002365/multiply-defined-symbols-mysqlclient-pthread-win32
+    # solve duplicate names from postgresql definitions
+    $(SED) -i 's/^void pthread_exit(/void mysql_pthread_exit(/g'  $(1)/include/my_pthread.h
+    $(SED) -i 's/^int pthread_cancel(/int mysql_pthread_cancel(/g'  $(1)/include/my_pthread.h
+    $(SED) -i 's/^int pthread_attr_init(/int mysql_pthread_attr_init(/g'  $(1)/include/my_pthread.h
+    $(SED) -i 's/^int pthread_attr_destroy(/int mysql_pthread_attr_destroy(/g'  $(1)/include/my_pthread.h
+    $(SED) -i 's/^int pthread_attr_getstacksize(/int mysql_pthread_attr_getstacksize(/g'  $(1)/include/my_pthread.h
+    $(SED) -i 's/^int pthread_attr_setstacksize(/int mysql_pthread_attr_setstacksize(/g'  $(1)/include/my_pthread.h
+    $(SED) -i 's/^int pthread_create(/int mysql_pthread_create(/g'  $(1)/include/my_pthread.h
+    $(SED) -i 's/^int pthread_join(/int mysql_pthread_join(/g'  $(1)/include/my_pthread.h
+    $(SED) -i 's/^void pthread_exit(/void mysql_pthread_exit(/g'  $(1)/mysys/my_winthread.c
+    $(SED) -i 's/^int pthread_cancel(/int mysql_pthread_cancel(/g'  $(1)/mysys/my_winthread.c
+    $(SED) -i 's/^int pthread_attr_init(/int mysql_pthread_attr_init(/g'  $(1)/mysys/my_winthread.c
+    $(SED) -i 's/^int pthread_attr_destroy(/int mysql_pthread_attr_destroy(/g'  $(1)/mysys/my_winthread.c
+    $(SED) -i 's/^int pthread_attr_getstacksize(/int mysql_pthread_attr_getstacksize(/g'  $(1)/mysys/my_winthread.c
+    $(SED) -i 's/^int pthread_attr_setstacksize(/int mysql_pthread_attr_setstacksize(/g'  $(1)/mysys/my_winthread.c
+    $(SED) -i 's/^int pthread_create(/int mysql_pthread_create(/g'  $(1)/mysys/my_winthread.c
+    $(SED) -i 's/^int pthread_join(/int mysql_pthread_join(/g'  $(1)/mysys/my_winthread.c
     # native build for tool comp_err
     # See https://bugs.mysql.com/bug.php?id=61340
     mkdir '$(1).native'

```

**Comment by tonytheodore** [comment]
@ByrAnd just apply the patches in https://github.com/mxe/mxe/pull/1009 or keep a branch of https://github.com/saiarcot895/mxe/tree/upstream-fix-posix-qt5 locally.

**Comment by tonytheodore** [comment]
fixed in #1488
